### PR TITLE
[WIP] Restructure Script generation

### PR DESCRIPTION
**Pull request by cloudRoutine**
- [x] Simplify and streamline the implementation
- [x] Disentangle implementations across dependency layers
- [x] Extract I/O execution out from generation module

### CODE REVIEW

File: src/Paket.Core/Installation/ScriptGeneration.fs [hunk 1]
```diff
@@ -1,467 +1,413 @@
...
+        
         let leafPackages =
-          openList 
-          |> List.filter (fun p ->
-              not (knownPackages.Contains(getPackageName p)) &&
-              getDependencies p |> Seq.forall (knownPackages.Contains))
+            openList |> List.filter (fun p ->
+                not (knownPackages.Contains(getPackageName p)) &&
+                getDependencies p |> Seq.forall (knownPackages.Contains)
+            )
```
Comment (matthid) [inline src/Paket.Core/Installation/ScriptGeneration.fs]: For some reason I like the code-style above more. But I'm not a code-style fanatic so I'm fine with this.

File: src/Paket.Core/Installation/ScriptGeneration.fs [hunk 2]
```diff
@@ -1,467 +1,413 @@
...
     let getDllOrder (dllFiles : AssemblyDefinition list) =
-      // this check saves looking at assembly metadata when we know this is not needed
-      if List.length dllFiles = 1 then
-          dllFiles 
-      else
-          // we ignore all unknown references as they are most likely resolved on package level
-          let known = dllFiles |> Seq.map (fun a -> a.FullName) |> Set.ofSeq
-          getPackageOrderGeneric
+        // this check saves looking at assembly metadata when we know this is not needed
+        if List.length dllFiles = 1 then dllFiles  else
```
Comment (matthid) [inline src/Paket.Core/Installation/ScriptGeneration.fs]: Wow that's really invisible, almost like a C# return statement :)

**Comment by smoothdeveloper** [comment]
I like the code duplication that went away (csharp & fsharp) and overall changes, although in general I like names which are a bit more descriptive (like `makeRelativePath` versus `relativePath` which sounds more like a variable, or `ScriptPiece` rather than just `Reference` which has very broad meaning in context of Paket).

There seems to be some extraneous changes not related to script generation, also `<|` spotted in a if.

**Comment by cloudRoutine** [comment]
`ResolveFrameworksForScriptGeneration`, `getFrameworkReferencesWithinPackage`, ``frameworksForDependencyGroups`

These kinds of names are excessive. documentation comments exist for a reason, all the information doesn't need to go into the binding's name 

it's not just a `Reference`, it's a `ScriptGeneration.Reference` that seems plenty clear to me.

```
if not (filterNuget <|string nuget) then
```
@smoothdeveloper you mean this? what about it?

**Comment by smoothdeveloper** [comment]
@cloudRoutine yes I meant that if:

`if not (filterNuget (string nuget))` or `if not <| filterNuget <| string nuget`, but the mix (and lack of space after the operator) makes it look a bit strange.

I understand your concern about kingdom of nouns, I think it becomes impediment where all symbols gets very long name difficult to distinguish from each others, and maybe you felt that with the code.

**Comment by cloudRoutine** [comment]
@smoothdeveloper the associativity on `<|` is garbage so it's not actually possible to use it like that




It's also why you see 
```fsharp
let (^) - (<|) 
```
in so much of my code, because that one does work 



regarding the names, it was nearly a wall of verbose and redundant text that made what should have been a very straightforward module very difficult to reason about.

**Comment by forki** [comment]
there is a merge conflict

**Comment by cloudRoutine** [comment]
It's still not done, working on it right now

File: Paket.sln [hunk 3]
```diff
@@ -1,18 +1,25 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
+VisualStudioVersion = 15.0.26403.7
+# Visual Studio 15
+VisualStudioVersion = 15.0.26228.10
```
Comment (matthid) [inline Paket.sln]: looks broken ;)

File: src/Paket.Core/Versioning/Requirements.fs [hunk 4]
```diff
@@ -503,12 +503,12 @@ let isTargetMatchingRestrictions =
     memoize <| fun (restrictions:FrameworkRestriction list, target) ->
         if List.isEmpty restrictions then true else
         match target with
-        | SinglePlatform (Runtimes _ ) -> true
+        //| SinglePlatform (Runtimes _ ) -> true
```
Comment (matthid) [inline src/Paket.Core/Versioning/Requirements.fs]: forgot to remove this...

File: src/Paket.Core/Versioning/PlatformMatching.fs [hunk 5]
```diff
@@ -173,14 +177,6 @@ let getTargetCondition (target:TargetProfile) =
...
         | DotNetCore(version) ->"$(TargetFrameworkIdentifier) == '.NETCoreApp'", sprintf "$(TargetFrameworkVersion) == '%O'" version
-        | DotNetUnity(version) when version = DotNetUnityVersion.V3_5_Full ->
-            "$(TargetFrameworkIdentifier) == '.NETFramework'", sprintf "($(TargetFrameworkVersion) == '%O' And $(TargetFrameworkProfile) == 'Unity Full v3.5')" version
-        | DotNetUnity(version) when version = DotNetUnityVersion.V3_5_Subset ->
-            "$(TargetFrameworkIdentifier) == '.NETFramework'", sprintf "($(TargetFrameworkVersion) == '%O' And $(TargetFrameworkProfile) == 'Unity Subset v3.5')" version
-        | DotNetUnity(version) when version = DotNetUnityVersion.V3_5_Micro ->
-            "$(TargetFrameworkIdentifier) == '.NETFramework'", sprintf "($(TargetFrameworkVersion) == '%O' And $(TargetFrameworkProfile) == 'Unity Micro v3.5')" version
-        | DotNetUnity(version) when version = DotNetUnityVersion.V3_5_Web ->
-            "$(TargetFrameworkIdentifier) == '.NETFramework'", sprintf "($(TargetFrameworkVersion) == '%O' And $(TargetFrameworkProfile) == 'Unity Web v3.5')" version
```
Comment (matthid) [inline src/Paket.Core/Versioning/PlatformMatching.fs]: can you add this again?

File: src/Paket.Core/PaketConfigFiles/ProjectFile.fs [hunk 6]
```diff
@@ -759,48 +754,40 @@ module ProjectFile =
...
+        // global targets are targets, that are either directly in the /build folder.
         // (ref https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#including-msbuild-props-and-targets-in-a-package). 
         let globalTargets, frameworkSpecificTargets =
             if not importTargets then List.empty, List.empty else
-            let sortedTargets = model.TargetsFileFolders |> List.sortBy (fun lib -> lib.Name)
+            let sortedTargets = model.TargetsFileFolders |> List.sortBy (fun lib -> lib.Path)
             sortedTargets
-            |> List.partition (fun lib -> (set lib.Targets).IsSupersetOf allTargetProfiles)
-        
+            |> List.partition (fun lib -> "" = lib.Path.Name)
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/ProjectFile.fs]: this was a mistake from my part...

File: src/Paket.Core/PaketConfigFiles/DependencyCache.fs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    let getPackageOrderResolvedPackage =
+        getPackageOrderGeneric 
+            (fun (p:PackageResolver.ResolvedPackage) -> p.Name) 
+            (fun p -> p.Dependencies |> Seq.map (fun (n,_,_) -> n))
+
+
+    //let getPackageOrderFromLockFile (lockFile:LockFile) =
+    //    lockFile.GetResolvedPackages ()
+    //    |> Seq.map (fun kvp -> kvp.Key , getPackageOrderResolvedPackage kvp.Value)
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/DependencyCache.fs]: can we remove this?

File: Paket.sln (hunk 3 above)
Comment (cloudRoutine) [inline Paket.sln]: hadn't noticed this one, `.sln` still opens fine 😆

File: src/Paket.Core/PackageAnalysis/FindReferences.fs [hunk 8]
```diff
@@ -4,7 +4,7 @@ open System
 open System.IO
 open Logging
 open Paket.Domain
-open Paket.InstallProcess
+//open Paket.InstallProcess
```
Comment (matthid) [inline src/Paket.Core/PackageAnalysis/FindReferences.fs]: remove?

File: src/Paket.Core/Installation/RestoreProcess.fs [hunk 9]
```diff
@@ -301,6 +304,8 @@ let Restore(dependenciesFileName,projectFile,force,group,referencesFileNames,ign
         |> Seq.toList
 
     if groupsToGenerate <> [] then
-        LoadingScripts.ScriptGeneration.executeCommand groupsToGenerate (DirectoryInfo dependenciesFile.RootPath) [] []
-
-    GarbageCollection.CleanUp(root, dependenciesFile, lockFile)
+        let depsCache = DependencyCache(dependenciesFile,lockFile) 
+        (LoadingScripts.ScriptGeneration.constructScriptsFromData depsCache groupsToGenerate [] [])
+        //(LoadingScripts.ScriptGeneration.constructScriptsFromDisk groupsToGenerate (DirectoryInfo dependenciesFile.RootPath) [] [])
```
Comment (matthid) [inline src/Paket.Core/Installation/RestoreProcess.fs]: remove?

File: RELEASE_NOTES.md [hunk 10]
```diff
@@ -1,5 +1,11 @@
-#### 4.6.3 - 24.04.2017
-* Unity3d version support - https://github.com/fsprojects/Paket/pull/2268
+#### 5.0.0-alpha002 - 19.04.2017
+* Fix dotnetcore integration.
+  * Paket now properly understands runtime and reference assemblies
+  * Paket now understands the runtime graph and restores runtime dependencies
+  * New API `InstallModel.GetRuntimeAssemblies` and `InstallModel.GetRuntimeLibraries` can be used to retrieve the correct assets for a particular RID and TFM
+* New command `paket generate-nuspec`
+* BREAKS: `InstallModel` API
```
Comment (matthid) [inline RELEASE_NOTES.md]: You might want to add your changes here

**Comment by matthid** [comment]
Obviously we need this to be green. You really seem to like red builds ;)

**Pull request merged**